File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_idx={}
        start=0
        mlen=0

        for end in range(len(s)):
            char=s[end]
            if char in char_idx:
                start=max(start,char_idx[char]+1)
            char_idx[char]=end
            mlen=max(mlen,end-start+1)

        return mlen

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int max=0,start=0;
        Map<Character,Integer>index=new HashMap<>();

        for(int i=0;i<s.length();i++){
            char c=s.charAt(i);
            if(index.containsKey(c)){
                start=start>index.get(c)+1?start:index.get(c)+1;
            }
            index.put(c,i);
            max=max>i-start+1?max:i-start+1;

        }
        return max;
    }
}